flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -131,340 +116,308 @@ class TimeBoundedStreamInnerJoin(
+      leftRow: Boolean): Unit = {
+    cRowWrapper.out = out
+    val row = cRowValue.row
+    if (!checkRowOutOfDate(timeForRow, myWatermark)) {
+      // Put the row into the cache for later use.
+      var rowList = rowListCache.get(timeForRow)
+      if (null == rowList) {
+        rowList = new ArrayList[Row](1)
+      }
+      rowList.add(row)
+      rowListCache.put(timeForRow, rowList)
+      // Register a timer on THE OPPOSITE STREAM to remove rows from the cache once they are
+      // expired.
+      if (oppositeTimeState.value == 0) {
+        registerCleanUpTimer(
+          ctx, timeForRow, oppositeWatermark, oppositeTimeState, leftRow, firstTimer = true)
+      }
 
-        // Join the record with records from the opposite stream.
-        val oppositeIterator = oppositeCache.iterator()
-        var oppositeEntry: Entry[Long, util.List[Row]] = null
-        var oppositeTime: Long = 0L;
-        while (oppositeIterator.hasNext) {
-          oppositeEntry = oppositeIterator.next
-          oppositeTime = oppositeEntry.getKey
-          if (oppositeTime < oppositeLowerBound - allowedLateness) {
-            //TODO Considering the data out-of-order, we should not remove records here.
-          } else if (oppositeTime >= oppositeLowerBound && oppositeTime <= oppositeUpperBound) {
-            val oppositeRows = oppositeEntry.getValue
-            var i = 0
-            if (leftRecord) {
-              while (i < oppositeRows.size) {
-                joinFunction.join(record, oppositeRows.get(i), cRowWrapper)
-                i += 1
-              }
-            } else {
-              while (i < oppositeRows.size) {
-                joinFunction.join(oppositeRows.get(i), record, cRowWrapper)
-                i += 1
-              }
+      // Join the row with rows from the opposite stream.
+      val oppositeIterator = oppositeCache.iterator()
+      while (oppositeIterator.hasNext) {
+        val oppositeEntry = oppositeIterator.next
+        val oppositeTime = oppositeEntry.getKey
+        if (oppositeTime >= oppositeLowerBound && oppositeTime <= oppositeUpperBound) {
+          val oppositeRows = oppositeEntry.getValue
+          var i = 0
+          if (leftRow) {
+            while (i < oppositeRows.size) {
+              joinFunction.join(row, oppositeRows.get(i), cRowWrapper)
+              i += 1
+            }
+          } else {
+            while (i < oppositeRows.size) {
+              joinFunction.join(oppositeRows.get(i), row, cRowWrapper)
+              i += 1
             }
-          } else if (oppositeTime > oppositeUpperBound) {
-            //TODO If the keys are ordered, can we break here?
           }
         }
-      } else {
-        //TODO Need some extra logic here?
-        LOG.warn(s"$record is out-of-date.")
+        // We could do the short-cutting optimization here once we get a state with ordered keys.
       }
     }
+    // We need to deal with the late data in the future.
```

Comment:
Totally agree. We don't need to handle late records. However, I think we should provide some runtime statistics about the late data. That could help the user know more about the data/the watermark assigner in use. What do you think?

Reply:
Yes, I agree. That would be a bigger effort though, because it should involve the other operators as well. So far we do not provide any metrics for the relational operators but it would make sense to do so.